src/binding/c/rma_api.txt

```diff
@@ -0,0 +1,466 @@
+# vim: set ft=c:
+
+MPI_Alloc_mem:
+    .desc: Allocate memory for message passing and RMA
+{
+    void *ap = MPID_Alloc_mem(size, info_ptr);
+
+    /* --BEGIN ERROR HANDLING-- */
+    if (!ap) {
+        mpi_errno =
+            MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
+                                 MPI_ERR_NO_MEM, "**allocmem", 0);
+        goto fn_fail;
+    }
+    /* --END ERROR HANDLING-- */
+
+    MPL_VG_MEM_INIT(ap, size);
+    *(void **) baseptr = ap;
+}
+
+MPI_Free_mem:
+    .desc: Free memory allocated with MPI_Alloc_mem
+{
+    if (base == NULL)
+        goto fn_exit;
+
+    mpi_errno = MPID_Free_mem(base);
+    if (mpi_errno)
+        goto fn_fail;
+}
+
+MPI_Accumulate:
+    .desc: Accumulate data into the target process using remote memory access
+    .skip: ThreadSafe
+    .seealso: MPI_Raccumulate
+    .impl: mpid
+    .earlyreturn: pt2pt_proc_null
+
+MPI_Compare_and_swap:
+    .desc: Perform one-sided atomic compare-and-swap.
+    .skip: ThreadSafe, validate-DATATYPE
+    .impl: mpid
+    .earlyreturn: pt2pt_proc_null
+/*
+    Notes:
+    This operation is atomic with respect to other "accumulate" operations.
+    
+    The parameter datatype must belong to one of the following categories of
+    predefined datatypes: C integer, Fortran integer, Logical, Multi-language
+    types, or Byte as specified in Section 5.9.2 on page 176. The origin and result
+    buffers (origin_addr and result_addr) must be disjoint.
+*/
+{ -- error_check --
```

Comment:
> Do we have a legitimate reason to check datatype when `count` is `0`? For example, is there standard text that implies e.g `MPI_Put(NULL, 0, MPI_DATATYPE_NULL, target, 0, 0, MPI_DATATYPE_NULL, win)` is wrong? On the other hand, adding special rules for RMA "userbuffer" check adds complexity to the python script that we should avoid unless necessary.

I don't think the above usage is a valid case in RMA. Unlike pt2pt, zero-count rma cannot be used as synchronization.  The only special case is fetch-like atomics with `NO_OP` where the origin_datatype can be `MPI_DATATYPE_NULL`. Thus, I don't think we need add the unnecessary  `if (origin_count > 0) {...}` code .

Reply:
> We have to manually annotate on which function need `PT2PT|COLL|RMA` specific logging and the rules from function names seems overly complex to add into the python scripts. Since I don't think there is anything specific about these logging -- probably from now removed `rlog` -- I took the simplified way.
> 
> Do you recall what was specific about these per-(pt2pt|coll|rma) logging?
> 
> The separate commit will essentially just rename the macros in those mpi functions that immediately get deleted in the next commit. If you insist, I'll do it.

In rlog, all are redirected to the same func. Maybe we can safely remove them. But should the change be in a sperate commit? 
```
#define RLOG_PT2PT_FUNC_ENTER(a)     RLOG_FUNC_ENTER(a)
#define RLOG_PT2PT_FUNC_EXIT(a)      RLOG_FUNC_EXIT(a)
#define RLOG_COLL_FUNC_ENTER(a)      RLOG_FUNC_ENTER(a)
#define RLOG_COLL_FUNC_EXIT(a)       RLOG_FUNC_EXIT(a)
#define RLOG_RMA_FUNC_ENTER(a)       RLOG_FUNC_ENTER(a)
#define RLOG_RMA_FUNC_EXIT(a)        RLOG_FUNC_EXIT(a)
```

Reply:
Ok, I will add the special condition for RMA in the python script.

Reply:
> > We have to manually annotate on which function need `PT2PT|COLL|RMA` specific logging and the rules from function names seems overly complex to add into the python scripts. Since I don't think there is anything specific about these logging -- probably from now removed `rlog` -- I took the simplified way.
> > Do you recall what was specific about these per-(pt2pt|coll|rma) logging?
> > The separate commit will essentially just rename the macros in those mpi functions that immediately get deleted in the next commit. If you insist, I'll do it.
> 
> In rlog, all are redirected to the same func. Maybe we can safely remove them. But should the change be in a sperate commit?
> 

Since there is zero impact removing the specific logging macros, I think we can skip the separation. The separation essentially is adding work that gets immediately removed, with no practical purpose at all.